include/autogradpp/detail.h

```diff
@@ -33,11 +35,80 @@ extern tag::Engine engine;
 class ContainerImpl;
 class OptimizerImpl;
 using Variable = tag::Variable;
-using variable_list = tag::variable_list;
 using Tensor = at::Tensor;
 using Container = std::shared_ptr<ContainerImpl>;
 using Optimizer = std::shared_ptr<OptimizerImpl>;
 
+#define GEN_TYPE(TYP, NAME) \
+  Variant(TYP);             \
+  bool is ## NAME () const; \
+  TYP get ## NAME () const;
+
+class Variant {
+ public:
+  Variant(Tensor);
+  Variant(Variable);
+  Variant(const std::string&);
+  Variant(std::vector<Variant>&);
+  Variant(std::vector<Variant>&&);
+  Variant(std::initializer_list<Variable>);
+  Variant(std::unordered_map<std::string, Variant>&);
+  Variant(std::unordered_map<std::string, Variant>&&);
```

Comment:
Maybe add initializer_list Variant? Or is that taken care of by the vector<Variant> constructors?